forc-test/src/execute.rs

```diff
@@ -23,10 +21,85 @@ use vm::interpreter::{InterpreterParams, MemoryInstance};
 use vm::state::DebugEval;
 use vm::state::ProgramState;
 
+#[derive(Debug, Clone)]
+pub enum CapturedEcal {
+    Write { fd: u64, bytes: Vec<u8> },
+    Unknown { ra: u64, rb: u64, rc: u64, rd: u64 },
+}
+
+impl CapturedEcal {
+    pub fn apply(&self) {
+        match self {
+            CapturedEcal::Write { fd, bytes } => {
+                let s = std::str::from_utf8(bytes.as_slice()).unwrap();
+
+                use std::io::Write;
+                use std::os::fd::FromRawFd;
+
+                let mut f = unsafe { std::fs::File::from_raw_fd(*fd as i32) };
+                write!(&mut f, "{}", s).unwrap();
+
+                // Dont close the fd
+                std::mem::forget(f);
+            }
+            CapturedEcal::Unknown { ra, rb, rc, rd } => {
+                println!("Unknown ecal: {} {} {} {}", ra, rb, rc, rd);
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct EcalState {
+    pub captured: Vec<CapturedEcal>,
+}
+
+impl EcalState {
+    pub fn clear(&mut self) {
+        self.captured.clear();
+    }
+}
+
+impl EcalHandler for EcalState {
+    fn ecal<M, S, Tx>(
+        vm: &mut Interpreter<M, S, Tx, Self>,
+        a: RegId,
+        b: RegId,
+        c: RegId,
+        d: RegId,
+    ) -> fuel_vm::error::SimpleResult<()>
+    where
+        M: fuel_vm::prelude::Memory,
+    {
+        let regs = vm.registers();
+        match regs[a.to_u8() as usize] {
+            1000 => {
```

Comment:
Should `1000` be put into some kind of consts file? This seems like a pretty important magic number that people should know about.

Reply:
Good idea.